### DISCUSSION TITLE
Getting Invalid Path when I pass a GPU device

### DISCUSSION THREAD

**Discussion by jayavanth** [cloud-hypervisor/cloud-hypervisor#5296] [Q&A]
**Describe the bug**
Getting Invalid Path when I pass a GPU device. I have a Tesla K40c and a GeForce GTX TITAN X. I get the same behavior with both GPUs

**To Reproduce**

I followed instruction from [custom-image.md](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/docs/custom-image.md#nvidia-image-for-vfio-baremetal-ci) and ran

```
./cloud-hypervisor-static -v \                         
        --kernel ./hypervisor-fw  \
        --disk path=jammy-server-cloudimg-amd64-nvidia.raw --disk path=/tmp/ubuntu-cloudinit.img \
        --cpus boot=4 \
        --memory size=4G \
        --net fd=3,mac=$mac 3<>"$tapdevice" \
       --device path="/sys/bus/pci/devices/0000:06:00.0" --device path="/sys/bus/pci/devices/0000:06:00.1"
```

I get this error `Error booting VM: VmBoot(DeviceManager(VfioCreate(InvalidPath)))`

This is my `lspci`

```
lspci | grep -iI nvidia
06:00.0 VGA compatible controller: NVIDIA Corporation GM200 [GeForce GTX TITAN X] (rev a1)
06:00.1 Audio device: NVIDIA Corporation GM200 High Definition Audio (rev a1)
0a:00.0 3D controller: NVIDIA Corporation GK180GL [Tesla K40c] (rev a1)
```

**Version**

Output of `cloud-hypervisor --version`: `cloud-hypervisor v30.0.0`

Did you build from source, if so build command line (e.g. features): No

**VM configuration**

Guest OS version details: From the `custom-image.md` example: `jammy-server-cloudimg-amd64-nvidia.raw`

Host OS version details: Ubuntu 20.04.4 LTS

**Logs**

Output of `cloud-hypervisor -v` from either standard error or via `--log-file`:
[gpu-path-not-found.log](https://github.com/cloud-hypervisor/cloud-hypervisor/files/11024773/gpu-path-not-found.log)

Linux kernel output:

**Comment by likebreath** [comment]
Did you unbind the devices from their native drivers and use the `vfio_pci` driver to manage them? You can double check that with `lspci -v`. 

Here is an example how to do that: https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/docs/vfio.md#example

**Comment by jayavanth** [comment]
Thank you for the documentation. I tried it and I see that it's being managed by `vfio_pci`

```
$ lspci -nnk | grep -A 2 K40c  
0a:00.0 3D controller [0302]: NVIDIA Corporation GK180GL [Tesla K40c] [10de:1024] (rev a1)
	Subsystem: NVIDIA Corporation GK180GL [Tesla K40c] [10de:0983]
	Kernel driver in use: vfio-pci
	Kernel modules: nvidiafb, nouveau, nvidia_drm, nvidia
```
Although when I run it now I get a different error. I'm guessing I'm getting close to getting it working

`Error booting VM: VmBoot(DeviceManager(VfioCreate(OpenGroup(Os { code: 13, kind: PermissionDenied, message: "Permission denied" }, "41"))))`
